utils/frame/benchmarking-cli/src/block/bench.rs

```diff
@@ -0,0 +1,187 @@
+		Self { client, params, _p: PhantomData }
+	}
+
+	/// Benchmark the execution speed of historic blocks and log the results.
+	pub fn run(&self) -> Result<()> {
+		if self.params.from == 0 {
+			return Err("Cannot benchmark the genesis block".into())
+		}
+
+		for i in self.params.from..=self.params.to {
+			let block_num = BlockId::Number(i.into());
+			let parent_num = BlockId::Number(((i - 1) as u32).into());
+			let consumed = self.consumed_weight(&block_num)?;
+
+			let block =
+				self.client.block(&block_num)?.ok_or(format!("Block {} not found", block_num))?;
+			let block = self.unsealed(block.block);
+			let took = self.measure_block(&block, &parent_num)?;
+
+			self.log_weight(block_num, block.extrinsics().len(), consumed, took);
+		}
+
+		Ok(())
+	}
+
+	/// Return the average *execution* aka. *import* time of the block.
+	fn measure_block(&self, block: &Block, parent_num: &BlockId<Block>) -> Result<NanoSeconds> {
+		let mut record = Vec::<NanoSeconds>::default();
+		// Interesting part here:
+		// Execute the block multiple times and collect stats about its execution time.
+		for _ in 0..self.params.repeat {
+			let block = block.clone();
+			let runtime_api = self.client.runtime_api();
+			let start = Instant::now();
+
+			runtime_api
+				.execute_block(&parent_num, block)
+				.map_err(|e| Error::Client(RuntimeApiError(e)))?;
+
+			record.push(start.elapsed().as_nanos() as NanoSeconds);
+		}
+
+		let took = Stats::new(&record)?.select(StatSelect::Average);
+		Ok(took)
+	}
+
+	/// Returns the total nanoseconds of a [`frame_system::ConsumedWeight`] for a block number.
+	///
+	/// This is the post-dispatch corrected weight and is only available
+	// after executing the block.
+	fn consumed_weight(&self, block: &BlockId<Block>) -> Result<NanoSeconds> {
+		// Hard-coded key for System::BlockWeight. It could also be passed in as argument
+		// for the benchmark, but I think this should work as well.
+		let hash = hex::decode("26aa394eea5630e07c48ae0c9558cef734abf5cb34d6244378cddbf18e849d96")?;
+		let key = StorageKey(hash);
+
+		let mut raw_weight = &self
+			.client
+			.storage(&block, &key)?
+			.ok_or(format!("Could not find System::BlockWeight for block: {}", block))?
+			.0[..];
+
+		let weight = ConsumedWeight::decode_all(&mut raw_weight)?;
+		// Should be divisible, but still use floats in case we ever change that.
+		Ok((weight.total() as f64 / WEIGHT_PER_NANOS as f64).floor() as NanoSeconds)
+	}
+
+	/// Prints the weight info of a block to the console.
+	fn log_weight(
+		&self,
+		num: BlockId<Block>,
+		num_ext: usize,
+		consumed: NanoSeconds,
+		took: NanoSeconds,
+	) {
+		// The ratio of weight that the block used vs the number that it consumed.
+		// This should in general not exceed 100% (minus outliers).
+		let percent = (took as f64 / consumed as f64) * 100.0;
+		let num = if let BlockId::Number(num) = num {
+			num
+		} else {
+			unreachable!("Only put numbers in; qed")
+		};
```

Comment:
Why then take `BlockId`?

Reply:
Ah yea, had the code structured differently before. Will change.